### PR TITLE
add fields such as CONTAINER_NAME to journald log entries sent to by containers

### PR DESCRIPTION
**Pull request by haytok**
In the current implementation, containers running by `nerdctl` don't 
export entries containing fileds such as `CONTAINER_NAME`, `IMAGE_NAME`
, and etc to the journald log like containers runnging by `docker cli`.

At this time, the journald log entry describes below when sending to the journald log using `nerdctl`.

<details><summary>Details</summary>
<p>

```
> nerdctl run -d --name nginx-nerdctl --log-driver=journald nginx
bb7df47d27fd73426cec286ed88c5abf1443e74df637e2440d2dbca7229a84dc

> nerdctl ps
CONTAINER ID    IMAGE                             COMMAND                   CREATED          STATUS    PORTS    NAMES
bb7df47d27fd    docker.io/library/nginx:latest    "/docker-entrypoint.…"    3 seconds ago    Up                 nginx-nerdctl

> sudo journalctl SYSLOG_IDENTIFIER=bb7df47d27fd -a -n 1 -o json-pretty
{
        "__CURSOR" : "???",
        "__REALTIME_TIMESTAMP" : "1730899940827182",
        "__MONOTONIC_TIMESTAMP" : "10815937979908",
        "_BOOT_ID" : "???",
        "_UID" : "0",
        "_GID" : "0",
        "_CAP_EFFECTIVE" : "1ffffffffff",
        "_MACHINE_ID" : "???",
        "_HOSTNAME" : "???.us-west-2.amazon.com",
        "_TRANSPORT" : "journal",
        "_SYSTEMD_SLICE" : "system.slice",
        "PRIORITY" : "3",
        "_SYSTEMD_CGROUP" : "/system.slice/containerd.service",
        "_SYSTEMD_UNIT" : "containerd.service",
        "_COMM" : "nerdctl",
        "_EXE" : "/usr/local/bin/nerdctl",
        "_CMDLINE" : "/usr/local/bin/nerdctl _NERDCTL_INTERNAL_LOGGING /var/lib/nerdctl/1935db59",
        "SYSLOG_IDENTIFIER" : "bb7df47d27fd",
        "_PID" : "8118",
        "MESSAGE" : "2024/11/06 13:32:20 [notice] 1#1: start worker process 44",
        "_SOURCE_REALTIME_TIMESTAMP" : "1730899940825905"
}
```

</p>
</details> 

On the other hand, the output fields are listed below when we use the journald logging driver with `docker cli`.

- https://docs.docker.com/engine/logging/drivers/journald/

As you can see, some entries are not output by `nerdctl` and are incompatible with `docker cli`.

This feature request is reported in the following:

- https://github.com/containerd/nerdctl/issues/3486

Therefore, in this pull request, we will add the fields to be output in 
the journald log.

After applying this fix, the journald log will output the following 
fields.

<details><summary>Details</summary>
<p>

```
> sudo journalctl SYSLOG_IDENTIFIER=fe22eccbd704 -a -n 1 -o json-pretty
{
  "__CURSOR": "???",
  "__REALTIME_TIMESTAMP": "1731385591671422",
  "__MONOTONIC_TIMESTAMP": "11301588824148",
  "_BOOT_ID": "???",
  "_MACHINE_ID": "???",
  "_HOSTNAME": "???.us-west-2.amazon.com",
  "PRIORITY": "3",
  "_TRANSPORT": "journal",
  "_UID": "0",
  "_GID": "0",
  "_COMM": "nerdctl",
  "_EXE": "/usr/local/bin/nerdctl",
  "_CMDLINE": "/usr/local/bin/nerdctl _NERDCTL_INTERNAL_LOGGING /var/lib/nerdctl/1935db59",
  "_CAP_EFFECTIVE": "1ffffffffff",
  "_SYSTEMD_CGROUP": "/system.slice/containerd.service",
  "_SYSTEMD_UNIT": "containerd.service",
  "_SYSTEMD_SLICE": "system.slice",
  "CONTAINER_NAME": "nginx-nerdctl",
  "IMAGE_NAME": "nginx",
  "CONTAINER_ID_FULL": "fe22eccbd704ba799785999079ac465ed067d5914e9e3f1020e769921d5a83c5",
  "SYSLOG_IDENTIFIER": "fe22eccbd704",
  "CONTAINER_TAG": "fe22eccbd704",
  "CONTAINER_ID": "fe22eccbd704",
  "_PID": "31643",
  "MESSAGE": "2024/11/12 04:26:31 [notice] 1#1: start worker process 44",
  "_SOURCE_REALTIME_TIMESTAMP": "1731385591669765"
}
```

</p>
</details> 

Note that details are described in each commits.

### CODE REVIEW

**Comment by haytok** [comment]
The CI faillures of `lint / go | windows` and `lint / go | linux` are related this PR.

- https://github.com/containerd/nerdctl/pull/3666

**Comment by haytok** [comment]
The following error in CI is related to this modification, so I'll investigate to resolve.

```
=== Failed
=== FAIL: cmd/nerdctl/container TestRunWithJournaldLogDriver/filter_using_IMAGE_NAME_field (20.00s)
    container_run_test.go:368: timeout hit after 20s: reading from journald is not yet finished

=== FAIL: cmd/nerdctl/container TestRunWithJournaldLogDriver (23.27s)

=== FAIL: cmd/nerdctl/container TestRunWithJournaldLogDriver/filter_using_IMAGE_NAME_field (re-run 1) (20.00s)
    container_run_test.go:368: timeout hit after 20s: reading from journald is not yet finished

=== FAIL: cmd/nerdctl/container TestRunWithJournaldLogDriver (re-run 1) (23.30s)

=== FAIL: cmd/nerdctl/container TestRunWithJournaldLogDriver/filter_using_IMAGE_NAME_field (re-run 2) (20.00s)
    container_run_test.go:368: timeout hit after 20s: reading from journald is not yet finished

=== FAIL: cmd/nerdctl/container TestRunWithJournaldLogDriver (re-run 2) (23.28s)

DONE 3 runs, 481 tests, 123 skipped, 6 failures in 767.586s
Error: Process completed with exit code 1.
```

**Comment by haytok** [comment]
After my investigation, I checked the docker cli doc and found no description of the details output in the `IMAGE_NAME` field.
However, it was confirmed that the current implementation (`nerdctl`) and the docker cli differ in the behaviour of the output to `IMAGE_NAME` in the image's Tag.

- https://docs.docker.com/engine/logging/drivers/journald/

> IMAGE_NAME	The name of the container image.

<details><summary>Details</summary>
<p>

In case of Docker

```
> CONTAINER_NAME="issue_3486"
> d run --rm --name $CONTAINER_NAME --log-driver=journald ghcr.io/stargz-containers/alpine:3.13-org sh -euxc "echo foo; echo bar"

+ echo foo
+ echo bar
foo
bar

> sudo journalctl CONTAINER_NAME=$CONTAINER_NAME -a -n 1 -o json-pretty | grep IMAGE_NAME
	"IMAGE_NAME" : "ghcr.io/stargz-containers/alpine:3.13-org",
```

In case of nerdctl in this modification.

```
> nerdctl run --rm --name $CONTAINER_NAME --log-driver=journald ghcr.io/stargz-containers/alpine:3.13-org sh -euxc "echo foo; echo bar"
+ echo foo
+ echofoo
bar
 bar

> sudo journalctl CONTAINER_NAME=$CONTAINER_NAME -a -n 1 -o json-pretty | grep IMAGE_NAME
	"IMAGE_NAME" : "ghcr.io/stargz-containers/alpine",
```

</p>
</details> 

Therefore, the `IMAGE_NAME` of the entry that nerdctl outputs to the journald log should be modified to be docker-compatible.

**Comment by haytok** [comment]
CI passed except for the lint issue which will be fixed in PR 3666.

**Comment by haytok** [comment]
@AkihiroSuda 

Thanks, I forgot to rebase ...
I rebased and pushed.

**Review by Zheaoli — APPROVED** [review]
LGTM 

BTW, I'm not sure whether we should let people config the log field they want

File: pkg/logging/journald_logger.go [hunk 1]
```diff
@@ -95,9 +101,38 @@ func (journaldLogger *JournaldLogger) PreProcess(dataStore string, config *loggi
 			syslogIdentifier = b.String()
 		}
 	}
+
+	ctx := context.Background()
```
Comment (fahedouch) [inline pkg/logging/journald_logger.go:105]: Please propagate the context to the `PreProcess` func

**Comment by haytok** [comment]
Hi @fahedouch 
I have fixed based on your review, so could you re-review when you have time?

File: pkg/logging/journald_logger.go [hunk 2]
```diff
@@ -200,3 +234,12 @@ func prepareJournalCtlDate(t string) (string, error) {
 	s := tm.Format("2006-01-02 15:04:05")
 	return s, nil
 }
+
+func GetJournaldImageNameField(image string) string {
+	imageName := image
+	if repo, tag := imgutil.ParseRepoTag(imageName); tag == "latest" {
```
Comment (fahedouch) [inline pkg/logging/journald_logger.go:240]: is it disturbing to to have a `latest` tag ?
 
I would rather check for empty tag using this [func](https://github.com/containerd/nerdctl/blob/main/pkg/imgutil/filtering.go#L352)

Comment (haytok) [reply pkg/logging/journald_logger.go:240]: Also, I'll add a test for the case where tag is latest.

Comment (haytok) [reply pkg/logging/journald_logger.go:240]: Thanks for the comments!!! @fahedouch 
When implementing tests, I noticed that Docker displays `the image name specified` when running a container in the IMAGE_NAME field ...

For example, we can check this behavior when running a container using the following image (`ghcr.io/containerd/busybox:latest`).

<details><summary>Details</summary>
<p>

```bash
> CONTAINER_NAME="latest-image"
> LATEST_IMAGE="ghcr.io/containerd/busybox:latest"

> d images | grep ghcr.io/containerd/busybox
ghcr.io/containerd/busybox         latest     16ea53ea7c65   3 years ago     1.24MB

> d run --rm --name $CONTAINER_NAME --log-driver=journald $LATEST_IMAGE sh -euxc "echo foo; echo bar"
+ echo foo
+ echo bar
foo
bar

> sudo journalctl -a CONTAINER_NAME=$CONTAINER_NAME -n 1 -o json-pretty | grep $LATEST_IMAGE
	"IMAGE_NAME" : "ghcr.io/containerd/busybox:latest",
```

</p>
</details> 

This behavior makes sense in the three cases above.
Therefore, I think it's not necessary to split the process according to the tag of image, so I'll fix the commit to delete `GetJournaldImageNameField()`.

File: pkg/logging/logging.go [hunk 3]
```diff
@@ -121,9 +121,11 @@ func Main(argv2 string) error {
...
 // LogConfig is marshalled as "log-config.json"
 type LogConfig struct {
-	Driver string            `json:"driver"`
-	Opts   map[string]string `json:"opts,omitempty"`
-	LogURI string            `json:"-"`
+	Driver      string            `json:"driver"`
+	Opts        map[string]string `json:"opts,omitempty"`
+	LogURI      string            `json:"-"`
+	Address     string            `json:"address"`
+	RowImageRef string            `json:"image"`
```
[the remark this answers is not in the document]
Comment (haytok) [reply pkg/logging/logging.go]: Thanks !!! Fixed.

File: pkg/logging/journald_logger.go [hunk 4]
```diff
@@ -95,9 +100,33 @@ func (journaldLogger *JournaldLogger) PreProcess(dataStore string, config *loggi
...
+	}
+
 	// construct log metadata for the container
 	vars := map[string]string{
 		"SYSLOG_IDENTIFIER": syslogIdentifier,
+		"CONTAINER_TAG":     syslogIdentifier,
+		"CONTAINER_ID":      shortID,
+		"CONTAINER_ID_FULL": containerID,
+		"CONTAINER_NAME":    containerutil.GetContainerName(containerLabels),
+		"IMAGE_NAME":        journaldLogger.RowImageRef,
```
[the remark this answers is not in the document]
Comment (haytok) [reply pkg/logging/journald_logger.go]: @fahedouch Thanks !!!

I think it is not possible to fetch row image ref (`journaldLogger.RawImageRef`) from the following `Labels()`, so I added `journaldLogger.RawImageRef`.

- https://github.com/containerd/containerd/blob/main/client/image.go#L48
- https://github.com/containerd/containerd/blob/main/client/container.go#L79

If I misunderstood your question, could you please tell me what `image ref` and `labels` are 🙇🙇

**Comment by haytok** [comment]
The following test has failed. I'm checking ...

<details><summary>Details</summary>
<p>

```bash
2024-11-25T06:17:46.8588148Z === RUN   TestRemove/Remove_image_with_paused_container_-_with_-f
2024-11-25T06:17:46.8589708Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 info --format {{ json . }}
2024-11-25T06:17:46.8591338Z     image_remove_test.go:306: ======================== Pre-test cleanup ========================
2024-11-25T06:17:46.8593845Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 rm -f testremove-remove-image-with-paused-container-with-f-41f5f718
2024-11-25T06:17:46.8595890Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 rmi -f 
2024-11-25T06:17:46.8597114Z     image_remove_test.go:306: ======================== Test setup ========================
2024-11-25T06:17:46.8600240Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 run --pull always -d --name testremove-remove-image-with-paused-container-with-f-41f5f718 ghcr.io/stargz-containers/alpine:3.13-org sleep 3600
2024-11-25T06:17:46.8603386Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 pause testremove-remove-image-with-paused-container-with-f-41f5f718
2024-11-25T06:17:46.8609064Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 image inspect ghcr.io/stargz-containers/alpine:3.13-org
2024-11-25T06:17:46.8611227Z     image_remove_test.go:306: ======================== Test Run ========================
2024-11-25T06:17:46.8613568Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 rmi -f ghcr.io/stargz-containers/alpine:3.13-org
2024-11-25T06:17:46.8615453Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 images
2024-11-25T06:17:46.8617060Z     command.go:112: assertion failed: expression is false: strings.Contains(stdout, compare): Output does not contain: "<none>"
2024-11-25T06:17:46.8618764Z         Command:  /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 images
2024-11-25T06:17:46.8619696Z         ExitCode: 0
2024-11-25T06:17:46.8620456Z         Stdout:   REPOSITORY    TAG    IMAGE ID    CREATED    PLATFORM    SIZE    BLOB SIZE
2024-11-25T06:17:46.8621312Z         
2024-11-25T06:17:46.8622809Z         Stderr:   time="2024-11-25T06:17:46Z" level=warning msg="content digest sha256:ec14c7992a97fc11425907e908340c6c3d6ff602f5f13d899e6b7027c9b4133a: not found"
2024-11-25T06:17:46.8624200Z         
2024-11-25T06:17:46.8624514Z         Env:
2024-11-25T06:17:46.8624877Z         HOSTNAME=ed64e3612023
2024-11-25T06:17:46.8625347Z         container=docker
2024-11-25T06:17:46.8625777Z         HOME=/root
2024-11-25T06:17:46.8626185Z         LANG=C.UTF-8
2024-11-25T06:17:46.8626669Z         INVOCATION_ID=7ddec279ffe441829a9ae2a4ba86fa56
2024-11-25T06:17:46.8627282Z         TERM=xterm
2024-11-25T06:17:46.8627666Z         SHLVL=3
2024-11-25T06:17:46.8628021Z         CGO_ENABLED=0
2024-11-25T06:17:46.8628442Z         _=/usr/local/bin/gotestsum
2024-11-25T06:17:46.8629455Z         PATH=/usr/local/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
2024-11-25T06:17:46.8653235Z         ***
2024-11-25T06:17:46.8654229Z         DOCKER_CONFIG=/tmp/TestRemoveRemove_image_with_paused_container_-_with_-f3820180234/001
2024-11-25T06:17:46.8655246Z         NERDCTL_TOML=/tmp/TestRemoveRemove_image_with_paused_container_-_with_-f3820180234/001/nerdctl.toml
2024-11-25T06:17:46.8656401Z     image_remove_test.go:306: ======================== Processing subtests ========================
2024-11-25T06:17:46.8657119Z     case.go:164: ======================== Post-test cleanup ========================
2024-11-25T06:17:46.8658348Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 rm -f testremove-remove-image-with-paused-container-with-f-41f5f718
2024-11-25T06:17:46.8659666Z     command.go:112: /usr/local/bin/nerdctl --namespace=testremove-private-a0b107b8 rmi -f ec14c799
2024-11-25T06:17:46.8660449Z --- FAIL: TestRemove/Remove_image_with_paused_container_-_with_-f (0.94s)
```

</p>
</details>

File: pkg/logging/journald_logger.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fahedouch) [reply pkg/logging/journald_logger.go]: I think it possible to fetch the image `ref` from [container.info(ctx)](https://github.com/containerd/containerd/blob/main/client/container.go#L56)

[the remark this answers is not in the document]
Comment (haytok) [reply pkg/logging/journald_logger.go]: Thank you, the image ref can be fetched from `container.Info(ctx)`, so I'll fix it.

**Comment by haytok** [comment]
At this time, the following test on Windows failed ...

- https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/image/image_remove_test.go#L116C1-L147C5

<details><summary>Details</summary>
<p>

```bash
2024-11-26T05:15:07.1142738Z === [31mFAIL[0m: cmd/nerdctl/image TestRemove/Remove_image_with_running_container_-_with_-f (6.94s)
2024-11-26T05:15:07.1143588Z     image_remove_test.go:306: ======================== Pre-test cleanup ========================
2024-11-26T05:15:07.1144904Z     command.go:112: C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 rm -f testremove-remove-image-with-running-container-with-f-e9c3b26f
2024-11-26T05:15:07.1146269Z     command.go:112: C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 rmi -f 
2024-11-26T05:15:07.1147090Z     image_remove_test.go:306: ======================== Test setup ========================
2024-11-26T05:15:07.1148872Z     command.go:112: C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 run --pull always -d --name testremove-remove-image-with-running-container-with-f-e9c3b26f gcr.io/k8s-staging-e2e-test-images/busybox:1.36.1-1 sleep 3600
2024-11-26T05:15:07.1150940Z     command.go:112: C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 image inspect gcr.io/k8s-staging-e2e-test-images/busybox:1.36.1-1
2024-11-26T05:15:07.1151994Z     image_remove_test.go:306: ======================== Test Run ========================
2024-11-26T05:15:07.1153186Z     command.go:112: C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 rmi -f gcr.io/k8s-staging-e2e-test-images/busybox:1.36.1-1
2024-11-26T05:15:07.1154447Z     command.go:112: C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 images
2024-11-26T05:15:07.1155529Z     command.go:112: assertion failed: expression is false: strings.Contains(stdout, compare): Output does not contain: "<none>"
2024-11-26T05:15:07.1156673Z         Command:  C:\Users\runneradmin\go\bin\nerdctl.exe --namespace=testremove-private-a0b107b8 images
2024-11-26T05:15:07.1157260Z         ExitCode: 0
2024-11-26T05:15:07.1157797Z         Stdout:   REPOSITORY    TAG    IMAGE ID    CREATED    PLATFORM    SIZE    BLOB SIZE
2024-11-26T05:15:07.1158268Z         
2024-11-26T05:15:07.1159289Z         Stderr:   time="2024-11-26T05:13:51Z" level=warning msg="content digest sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9: not found"
2024-11-26T05:15:07.1160210Z         
```

</p>
</details> 

When we try to `nerdctl rmi -f` the image that the container is running, a none image should be left with.
I'm checking ...

**Comment by haytok** [comment]
Hi, @fahedouch 

All tests have passed, so could you check when you have time?

**Pull request merged**